### PR TITLE
AT: implement linting AT generation result

### PR DESCRIPTION
**Pull request by SelyankinFyodor**
-separate creating, linting and testing templates.
-add new npm scripts
-minor refactoring

### CODE REVIEW

File: packages/devextreme-cli/testing/creating.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+const vueEnv = require('./env.vue');
+const angularEnv = require('./env.angular');
+const minimist = require('minimist');
+
+const args = minimist(process.argv.slice(), {
+    default: {
+        template: 'all'
+    },
+    alias: {
+        t: 'template'
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/creating.js:11]: Why this argument named so? Previously it was a [testNamePattern](https://jestjs.io/docs/en/cli#--testnamepatternregex)  flag that was passed to `jest`. It looks like it has another meaning now.

File: packages/devextreme-cli/testing/app-template.test.shared.js [hunk 2]
```diff
@@ -8,7 +8,7 @@ const DevServer = require('./dev-server');
 const defaultLayout = 'side-nav-outer-toolbar';
 
 module.exports = (env) => {
-    const skipAppCreation = process.env.TEST_MODE && process.env.TEST_MODE === 'dev';
+    console.log('snapshot TESTING: ' + env.engine);
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/app-template.test.shared.js:11]: Why do we need this `console.log` here?

File: packages/devextreme-cli/testing/creating.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+const envs = {
+    'react': reactEnv,
+    'angular': angularEnv,
+    'vue': vueEnv
+};
+
+(async function createApp() {
+    if(!(args.t in envs)) {
+        console.log(args.t);
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/creating.js:23]: Consoles again. Do we need it?

File: packages/devextreme-cli/testing/dev-server.js [hunk 4]
```diff
@@ -2,6 +2,8 @@ const { spawn } = require('child_process');
 const fs = require('fs');
 const path = require('path');
 const kill = require('tree-kill');
+// import kill from 'tree-kill-promise';
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/dev-server.js:5]: Definitely it's not useful changes :)

File: packages/devextreme-cli/testing/env.angular.js [hunk 5]
```diff
@@ -31,12 +31,13 @@ exports.appPath = appPath;
 exports.npmArgs = ['start', '--', '--host', '0.0.0.0'];
 
 exports.createApp = async() => {
+    console.log('START CREATE ANGULAR APP');
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/env.angular.js:34]: WHY?

File: packages/devextreme-cli/testing/lint-config/angular.eslintrc [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+        "ecmaVersion": 2018,
+        "sourceType": "module"
+    },
+    "env": {
+        "node": true
+    },
+    "rules": {
+        "angular/log": 0
+    }
+}
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/lint-config/angular.eslintrc:14]: https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file

File: packages/devextreme-cli/testing/linting.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
+const reactEnv = require('./env.react');
+const vueEnv = require('./env.vue');
+const angularEnv = require('./env.angular');
+const minimist = require('minimist');
+const fs = require('fs');
+const ESLint = require('eslint').ESLint;
+
+const args = minimist(process.argv.slice(), {
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/linting.js:8]: It looks like a duplication of code. Maybe it's better to extract some utility that will get an array of engines and filter it by the argument? Something like this:
```js
const environments = [
  require('./env.react'),
  require('./env.vue'),
  require('./env.angular')
];
filterEnvs(environments, provess.argv).forEach((env) => {
  ...
})
```

File: packages/devextreme-cli/testing/linting.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+});
+
+const envs = {
+    'react': reactEnv,
+    'angular': angularEnv,
+    'vue': vueEnv
+};
+
+async function lint(env) {
+    console.log('START LINT: ' + env.engine);
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/linting.js:24]: Are all of these console.log calls useful?

File: packages/devextreme-cli/testing/linting.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    let eslint = new ESLint({
+        useEslintrc: false,
+        overrideConfigFile: `./testing/lint-config/${env.engine}.eslintrc`,
+        ignore: false
+    });
+    let report = await eslint.lintFiles([
+        `./testing/sandbox/${env.engine}/my-app/src/**/*.${env.engine === 'angular' ? 'ts' : 'js'}`
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/linting.js:31]: Maybe it's better to store the information about file's extension right in env files?

File: packages/devextreme-cli/testing/linting.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+};
+
+
+(async function lintProcess() {
+    if(!(args.t in envs)) {
+        Object.keys(envs).forEach(async env => {
+            if(fs.existsSync(envs[env].appPath)) {
+                await lint(envs[env]);
+            } else {
+                console.log(`!!!!!!! ${envs[env].appPath} IS NOT EXIST!!!!!!!`);
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/linting.js:49]: Should not we throw an error in this case?

File: packages/devextreme-cli/testing/testing.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        silent: false,
+        forceExit: true,
+        showConfig: false,
+        detectOpenHandles: true,
+        runInBand: true
+    };
+
+    await jest
+        .runCLI(options, options.projects).then(
+            resolve => {},
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/testing.js:59]: It looks strange. Why is that for?

File: packages/devextreme-cli/testing/testing.js [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    }
+});
+
+const envs = {
+    'react': reactEnv,
+    'angular': angularEnv,
+    'vue': vueEnv
+};
+
+let killerEvent = new Emmiter();
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/testing.js:25]: Why not `const`?

File: packages/devextreme-cli/testing/testing.js [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+let done = Object.keys(envs).length;
+killerEvent.on('kill', function() {
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/testing.js:27]: Why do we need separate event for this. What if it will be a method with the same `done` counter?

**Review by nightskylark — CHANGES_REQUESTED** [review]

File: packages/devextreme-cli/testing/linting.js (hunk 9 above)
Comment (SelyankinFyodor) [reply packages/devextreme-cli/testing/linting.js:31]: `eslint.lintFiles` requires a string or array of string. Tried adding the file pattern to the config file. as a result, skips linting.

Comment (SelyankinFyodor) [reply packages/devextreme-cli/testing/linting.js:31]: corrected to a shorter entry

File: packages/devextreme-cli/testing/linting.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    let filteredEnvs = envs.filter(e => e.engine === args.e);
+    if(!filteredEnvs.length && args.e === 'all') {
+        filteredEnvs = envs;
+    }
+    filteredEnvs.forEach(async env => {
+        if(fs.existsSync(env.appPath)) {
+            await lint(env);
+        }
+    });
+})().catch(reject => console.error('\x1b[31m%s\x1b[0m', reject));
```
Comment (nightskylark) [inline packages/devextreme-cli/testing/linting.js:55]: ''\x1b[31m%s\x1b[0m''
WTF?

**Review by nightskylark — CHANGES_REQUESTED** [review]

**Review by nightskylark — APPROVED** [review]

**Review by LazyLahtak — APPROVED** [review]

**Review by nightskylark — APPROVED** [review]

**Pull request merged**